### ISSUE TITLE
A `validate=False` an option for `graph_objects` and `px` Figures?

### ISSUE THREAD

**Issue by michaelbabyn** [plotly/plotly.py#1812]
There's already an [issue outlining the effects `graph_object` validation has on plot generation time](https://github.com/plotly/plotly.py/issues/1052). Users can bypass this performance hit by replacing the `graph_objects` with `dict` and then display the plot with `plotly.offline.iplot(fig, validate=False)` or if they are creating graphs in Dash, they can forgo the `plotly.py` library altogether and just use a `dict` in their Graph component's `figure` argument. 

This solution can [greatly improve the performance of Dash apps](https://github.com/plotly/dash-sample-apps/pull/332) but it means that Dash users with expensive graphs have to choose between using `px`/plotly.py's `update` methods and optimally fast code. 

I wonder if a way to turn off validation, especially in Dash apps, would help Dash users get the best of both worlds.

cc @matthewchan15

**Comment by emmanuelle** [comment]
Also related to https://community.plot.ly/t/plotting-large-number-of-graphs/35907.

**Comment by emmanuelle** [comment]
To be checked: can we do this and still keep the magical underscore methods? 

Also possible: half-way point where we would disable the validation of only data arrays.

Note that the "import" time is a big part of the lag when developing

**Comment by parksj10** [comment]
Any update on this? certainly have my +1, using large data sets with datashader and it's taking seconds to validate. Likely will have to retrofit my code with the dict methods :(

**Comment by nicolaskruchten** [comment]
@parksj10 can you confirm you’re seeing performance issues with a version of plotly of 4.7 or higher? We made a number of performance improvements in 4.7 so I just want to make sure :)

**Comment by parksj10** [comment]
@nicolaskruchten running plotly 4.8.1, I've attached a cProfile below, you can see that half the figure generation time is spent validating. In case you're interested, I've also attached the cProfile .dat file. Let me know if I can do anything else to help or provide other information. I think it would be rather difficult to create a low-complexity, working example from my app, but perhaps @michaelbabyn 's examples could be useful in this regard

![Screen Shot 2020-06-25 at 7 41 04 PM](https://user-images.githubusercontent.com/28692761/85815021-3686e980-b71c-11ea-96a7-da09a5493701.png)

[temp.dat.zip](https://github.com/plotly/plotly.py/files/4834815/temp.dat.zip)

**Comment by nicolaskruchten** [comment]
Thanks! This is something we should fix, and we’d appreciate any help :)

**Comment by ndrezn** [comment]
Also am running into this. This causes major issues when working with e.g. choropleth maps with large GeoJSON files, where you will end up with giant JSON blobs that certainly do not need to be validated. 

I imagine this is a pretty common issue for folks working with charts with many points, and I had no idea this was even a thing until today. It'd be great at least to document this behaviour or make people more aware of it until it's possible to disable validation. Maybe even on https://plotly.com/python/webgl-vs-svg/?

**Comment by alexcjohnson** [comment]
I like the idea of a three-level approach: full validation (current behavior), top-level validation (don’t dig into data arrays or nested objects like GeoJSON), and no validation.

**Comment by ndrezn** [comment]
(want to note as well that I'm seeing ~1second validation time/mb of object. With GeoJSONs, we often see blobs in the size of 60mb+, which just destroys your app performance)

**Comment by nicolaskruchten** [comment]
So independently of the validation issue, if the GeoJSONs are static, you should always load them from `assets` in a Dash app, for caching purposes. Basically just pass in the URL rather than the GeoJSON blob.

**Comment by nicolaskruchten** [comment]
> Having the top-level validation option seems perfect!

Yes, of course, although the last time we tried, we were unable to make it work :)

**Comment by ndrezn** [comment]
(cc @red-patience / @LiamConnors on that last point maybe)

**Comment by hannahker** [comment]
Throwing my support behind this one! Even if it takes some time to add in a `validate=False` param, in the meantime it would be really helpful to have documentation to alert people that this might be a bottleneck in chart performance and that you can work around it with creating the `dict` directly. 

Both this trick and passing data as a static `asset` url have **massively** improved the performance of my graph and I wouldn't have known to do either of these things if I hadn't been pointed towards this issue. 

cc @red-patience

**Comment by bmaranville** [comment]
I think I have a related issue affecting `subplots.make_subplots`, where the time to execute increases non-linearly with the number of plots.  For a 20x20 grid of plots it is taking 14 seconds, for a 21x21 grid it takes 18 seconds, for example.  This is for an *empty* figure, which is created with make_subplots e.g. 
```python
from plotly.subplots import make_subplots

%time fig = make_subplots(rows=20, cols=20)
```

From profiling, it is spending the vast majority of its time in the `_ret` function of basedatatypes.py, and all of the time in that function is spent in `find_closest_string`, which I think is because it is pre-calculating an error message for a missing key - which is related to the validation.  There would be a > 90% speedup if validation could be disabled, from what I can see in the profiling.

**Comment by nicolaskruchten** [comment]
Thanks for that profiling! We could probably speed things up by only computing error strings when we know there's an error...

**Comment by AlonSpivack** [comment]
I've filed a detailed proposal in #5514 that addresses this issue with a different approach: 
an `_as_dict=True` flag on both `go.Figure` and graph object constructors.

The existing `_validate=False` parameter only provides ~5x speedup (892 → 293 function calls) 
because `__init__` still allocates the full object hierarchy and loops through all properties. 
The `_as_dict=True` approach intercepts in `__new__` before `__init__` ever runs, achieving 
~300x speedup for trace construction (6 function calls) and 26x for end-to-end figure pipelines.

I've prototyped all changes in `basedatatypes.py` - no codegen changes needed - and existing 
tests pass. I'm ready to open a PR if maintainers are interested.

See #5514 for full benchmarks, implementation details, and correctness verification.
